rasa/utils/tensorflow/data_generator.py

```diff
@@ -0,0 +1,428 @@
+    def _filter_out_fake_inputs(
+        array_of_array_of_features: FeatureArray,
+    ) -> Union[List[List[np.ndarray]], List[List[scipy.sparse.spmatrix]]]:
+        return list(
+            filter(
+                # filter empty lists created by another filter
+                lambda x: len(x) > 0,
+                [
+                    # filter all the "fake" inputs, we know the input is "fake",
+                    # when sequence dimension is `0`
+                    list(filter(lambda x: x.shape[0] > 0, array_of_features))
+                    for array_of_features in array_of_array_of_features
+                ],
+            )
+        )
+
+
+class RasaBatchDataGenerator(RasaDataGenerator):
+    """Data generator with an optional increasing batch size."""
+
+    def __init__(
+        self,
+        model_data: RasaModelData,
+        batch_size: Union[List[int], int],
+        epochs: int = 1,
+        batch_strategy: Text = SEQUENCE,
+        shuffle: bool = True,
+    ):
+        """Initializes the increasing batch size data generator.
+
+        Args:
+            model_data: The model data to use.
+            batch_size: The batch size.
+            epochs: The total number of epochs.
+            batch_strategy: The batch strategy.
+            shuffle: If 'True', data will be shuffled.
+        """
+        self.increase_batch_size = False
+        if isinstance(batch_size, list):
+            logger.debug(
+                "The provided batch size is a list, this data generator will use a "
+                "linear increasing batch size."
+            )
+            self.increase_batch_size = True
+        self.epochs = epochs
+        self.current_epoch = -1
+        self.current_batch_size = batch_size
+
+        super().__init__(model_data, batch_size, batch_strategy, shuffle)
+
+    def __len__(self) -> int:
+        """Number of batches in the Sequence.
+
+        Returns:
+            The number of batches in the Sequence.
+        """
+        num_examples = self.model_data.num_examples
+        batch_size = self.current_batch_size
+        return num_examples // batch_size + int(num_examples % batch_size > 0)
+
+    def __getitem__(self, index: int) -> Tuple[Any, Any]:
+        """Gets batch at position `index`.
+
+        Arguments:
+            index: position of the batch in the Sequence.
+
+        Returns:
+            A batch (tuple of input data and target data).
+        """
+        start = index * self.current_batch_size
+        end = start + self.current_batch_size
+
+        # return input and target data, as our target data is inside the input
+        # data return None for the target data
+        return self.prepare_batch(self.model_data.data, start, end), None
+
+    def on_epoch_end(self) -> None:
+        """Update the data after every epoch."""
+        self.current_epoch += 1
+        if self.increase_batch_size:
+            self.current_batch_size = self.linearly_increasing_batch_size(
```

Comment:
Probably doesn't matter in terms of performance but to replicate the results of master exactly, the call to `self.linearly_increasing_batch_size` should be inside `on_epoch_begin` instead of `on_epoch_end`.

Reply:
yeah, but they don't seem to have `on_epoch_begin` in keras

Reply:
looking at it, it seems to be equivalent to calling on the first epoch, except the call in init

Reply:
slightly refactored, to make it a bit more clear